akka-stream/src/main/scala/akka/stream/impl/ActorRefSource.scala

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.actor.{ ActorRef, PoisonPill }
+import akka.annotation.InternalApi
+import akka.stream.OverflowStrategies._
+import akka.stream._
+import akka.stream.stage._
+import akka.util.OptionVal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object ActorRefSource {
+  case object EagerComplete
+
+  private sealed trait ActorRefStage { def ref: ActorRef }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class ActorRefSource[T](
+    maxBuffer: Int,
+    overflowStrategy: OverflowStrategy,
+    completionMatcher: PartialFunction[Any, Unit],
+    failureMatcher: PartialFunction[Any, Throwable])
+    extends GraphStageWithMaterializedValue[SourceShape[T], ActorRef] {
+  import ActorRefSource._
+
+  val out: Outlet[T] = Outlet[T]("actorRefSource.out")
+
+  override val shape: SourceShape[T] = SourceShape.of(out)
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ActorRef) =
+    throw new IllegalStateException("Not supported")
+
+  private[akka] override def createLogicAndMaterializedValue(
+      inheritedAttributes: Attributes,
+      eagerMaterializer: Materializer): (GraphStageLogic, ActorRef) = {
+    val stage: GraphStageLogic with StageLogging with ActorRefStage = new GraphStageLogic(shape) with StageLogging
+    with ActorRefStage {
+      override protected def logSource: Class[_] = classOf[ActorRefSource[_]]
+
+      private val buffer: OptionVal[Buffer[T]] =
+        if (maxBuffer != 0)
+          OptionVal(Buffer(maxBuffer, eagerMaterializer))
+        else {
+          OptionVal.None // for backwards compatibility with old actor publisher based implementation
+        }
+      private var isCompleting: Boolean = false
+
+      override def preStart(): Unit = {
+        super.preStart()
+        if (maxBuffer == 0)
+          log.warning("for backwards compatibility: maxBuffer of 0 will not be supported in the future") // warning for backwards compatibility
+      }
+
+      override protected def stageActorName: String =
+        inheritedAttributes.get[Attributes.Name].map(_.n).getOrElse(super.stageActorName)
+
+      val ref: ActorRef = getEagerStageActor(eagerMaterializer, poisonPillFallback = true) {
+        case (_, EagerComplete) ⇒
+          completeStage()
+        case (_, PoisonPill) ⇒
+          log.warning("for backwards compatibility: PoisonPill will note be supported in the future")
+          isCompleting = true
+          tryPush()
```

Comment:
You're right, `PoisonPill` should immediately complete. I've added the `EagerComplete` as a replacement for it in my initial proposal.
Another option would be (when we break it and remove `PoisonPill`) to extend the `completionMatcher` to return something else instead of `Unit` to indicate if it should complete eager or not.

Reply:
I've pushed a proposal for replacing `PoisonPill` in the future.

Reply:
Interestingly the Java API doc contrary to the Scala API doc warns from using `PoisonPill`: 

``   * Note that terminating the actor without first completing it, either with a success or a
   * failure, will prevent the actor triggering downstream completion and the stream will continue
   * to run even though the source actor is dead. Therefore you should **not** attempt to
   * manually terminate the actor such as with a [[akka.actor.PoisonPill]]. ``